File: backend/app/schemas/datasets.py
from typing import Any, Literal

from pydantic import BaseModel, Field


class CoordinateColumns(BaseModel):
    x: str
    y: str
    z: str | None = None


class MeshMetadata(BaseModel):
    available: bool = True
    source: Literal["original"] = "original"
    node_count: int = Field(alias="nodeCount")
    element_count: int = Field(alias="elementCount")
    element_type_counts: dict[str, int] = Field(alias="elementTypeCounts")
    domain_count: int | None = Field(default=None, alias="domainCount")
    boundary_count: int | None = Field(default=None, alias="boundaryCount")
    mesh_dimension: Literal["1D", "2D", "3D"] = Field(alias="meshDimension")
    node_fields: list[str] = Field(default_factory=list, alias="nodeFields")
    element_fields: list[str] = Field(default_factory=list, alias="elementFields")
    visualization_tier: Literal["full", "surface", "decimated", "progressive"] = Field(
        alias="visualizationTier"
    )


class DatasetMetadata(BaseModel):
    dataset_id: str = Field(alias="datasetId")
    filename: str
    row_count: int = Field(alias="rowCount")
    dimension: Literal["2D", "3D"]
    coordinate_columns: CoordinateColumns = Field(alias="coordinateColumns")
    fields: list[str]
    bounds: dict[str, tuple[float, float]]
    mesh: MeshMetadata | None = None


class PointCoordinates(BaseModel):
    x: list[float]
    y: list[float]
    z: list[float] | None = None


class PointDataResponse(BaseModel):
    dataset_id: str = Field(alias="datasetId")
    total_points: int = Field(alias="totalPoints")
    returned_points: int = Field(alias="returnedPoints")
    downsampled: bool
    row_indexes: list[int] = Field(alias="rowIndexes")
    coordinates: PointCoordinates
    fields: dict[str, list[float | None]]


class MeshCoordinates(BaseModel):
    x: list[float]
    y: list[float]
    z: list[float]


class MeshDataResponse(BaseModel):
    dataset_id: str = Field(alias="datasetId")
    source: Literal["original"] = "original"
    node_ids: list[str] = Field(alias="nodeIds")
    coordinates: MeshCoordinates
    node_fields: dict[str, list[float | None]] = Field(alias="nodeFields")
    element_ids: list[str] = Field(alias="elementIds")
    element_types: list[str] = Field(alias="elementTypes")
    element_offsets: list[int] = Field(alias="elementOffsets")
    connectivity: list[int]
    domain_ids: list[str | None] = Field(alias="domainIds")
    boundary_ids: list[str | None] = Field(alias="boundaryIds")
    element_fields: dict[str, list[float | None]] = Field(alias="elementFields")
    surface_triangles: list[int] = Field(alias="surfaceTriangles")
    surface_owners: list[int] = Field(alias="surfaceOwners")
    surface_edges: list[int] = Field(alias="surfaceEdges")
    all_edges: list[int] = Field(alias="allEdges")
    statistics: MeshMetadata


class ActionRequest(BaseModel):
    action: Literal[
        "find_max",
        "find_min",
        "statistics",
        "filter",
        "profile",
        "nearest_point",
    ]
    params: dict[str, Any] = Field(default_factory=dict)

File: frontend/types/datasets.ts
export type Dimension = "2D" | "3D";
export type Representation = "surface" | "points" | "wireframe";
export type VisualizationMode = "field" | "surface" | "slice" | "isosurface" | "streamlines" | "vector" | "mesh";
export type MeshRenderType = "wireframe" | "surface" | "volume" | "field";
export type ThresholdOperator = ">" | ">=" | "<" | "<=" | "==";

export interface CoordinateColumns {
  x: string;
  y: string;
  z?: string | null;
}

export interface DatasetMetadata {
  datasetId: string;
  filename: string;
  rowCount: number;
  dimension: Dimension;
  coordinateColumns: CoordinateColumns;
  fields: string[];
  bounds: Record<string, [number, number]>;
  mesh?: MeshMetadata | null;
}

export interface MeshMetadata {
  available: true;
  source: "original";
  nodeCount: number;
  elementCount: number;
  elementTypeCounts: Record<string, number>;
  domainCount?: number | null;
  boundaryCount?: number | null;
  meshDimension: "1D" | "2D" | "3D";
  nodeFields: string[];
  elementFields: string[];
  visualizationTier: "full" | "surface" | "decimated" | "progressive";
}

export interface MeshData {
  datasetId: string;
  source: "original";
  nodeIds: string[];
  coordinates: { x: number[]; y: number[]; z: number[] };
  nodeFields: Record<string, Array<number | null>>;
  elementIds: string[];
  elementTypes: string[];
  elementOffsets: number[];
  connectivity: number[];
  domainIds: Array<string | null>;
  boundaryIds: Array<string | null>;
  elementFields: Record<string, Array<number | null>>;
  surfaceTriangles: number[];
  surfaceOwners: number[];
  surfaceEdges: number[];
  allEdges: number[];
  statistics: MeshMetadata;
}

export interface MeshSettings {
  renderType: MeshRenderType;
  showMesh: boolean;
  showSurfaceElements: boolean;
  showInternalElements: boolean;
  showEdges: boolean;
  showNodes: boolean;
  opacity: number;
  edgeThickness: number;
  nodeSize: number;
  elementColoring: boolean;
  clip: { x: number | null; y: number | null; z: number | null };
  sliceAxis: "xy" | "xz" | "yz" | "custom";
  customNormal: { x: number; y: number; z: number };
  customPosition: number;
  selectionMode: "element" | "node";
  projection: "perspective" | "orthographic";
  quality: "auto" | "high" | "balanced";
  isoValue: number | null;
  vectorScale: number;
}

export type MeshSelection =
  | {
      kind: "node";
      nodeIndex: number;
      nodeId: string;
      location: { x: number; y: number; z: number };
      values: Record<string, number>;
    }
  | {
      kind: "element";
      elementIndex: number;
      elementId: string;
      elementType: string;
      nodeIds: string[];
      domainId: string | null;
      boundaryId: string | null;
      centroid: { x: number; y: number; z: number };
      values: Record<string, number>;
    };

export interface PointData {
  datasetId: string;
  totalPoints: number;
  returnedPoints: number;
  downsampled: boolean;
  rowIndexes: number[];
  coordinates: {
    x: number[];
    y: number[];
    z?: number[] | null;
  };
  fields: Record<string, Array<number | null>>;
}

export interface Threshold {
  field: string;
  operator: ThresholdOperator;
  value: number;
}

export interface SelectedPoint {
  rowIndex: number;
  location: { x: number; y: number; z?: number };
  values: Record<string, number>;
}

export interface StatisticsResult {
  action: "statistics";
  field: string;
  count: number;
  min: number;
  max: number;
  mean: number;
  median: number;
  standardDeviation: number;
  range: number;
  minLocation: { x: number; y: number; z?: number };
  maxLocation: { x: number; y: number; z?: number };
}

export interface ExtremeResult {
  action: "find_max" | "find_min";
  field: string;
  value: number;
  rowIndex: number;
  location: { x: number; y: number; z?: number };
}

export interface FilterResult {
  action: "filter";
  field: string;
  operator: ThresholdOperator;
  value: number;
  matchedCount: number;
  returnedCount: number;
  truncated: boolean;
  rowIndexes: number[];
}

export interface ProfileResult {
  action: "profile";
  field: string;
  axis: "x" | "y" | "z";
  points: Array<{ position: number; value: number }>;
}

export interface NearestPointResult extends SelectedPoint {
  action: "nearest_point";
}

export interface VisualizationState {
  activeField: string;
  threshold: Threshold | null;
  selectedPoint: SelectedPoint | null;
  representation: Representation;
}
